### DISCUSSION TITLE
Help needed: First out of memory, now failed to create thumb

### DISCUSSION THREAD

**Discussion by boldeuler** [pocketbase/pocketbase#5925] [Q&A]
```
Pocketbase version: [PocketBase v0.22.26]
```
My pocketbase server is running on Ubuntu 24 behind  proxy and was not delivering images last night.

While checking the Logs I noticed 

```
unable to open database file: out of memory (14); failed query: SELECT {{_collections}}.* FROM _collections WHERE [[id]] = {:id} OR LOWER([[name]])={:name} LIMIT 1
```

I have 64GB RAM, and enough SSD (1TB) and HDD space (48TB).

data.db and logs.db are on the SSD, storage is on the HDD.

Side Note: the size of logs.db is 17GB

After restarting using systemctl restart pocketbase, it worked again but now it fails to generate thumbs, the logs show

```
Fallback to original - failed to create thumb 0x360
```

This is my pocketbase.service

```
[Unit]
Description=PocketBase Service
After=network.target

[Service]

Type           = simple
User           = root
Group          = root
LimitNOFILE    = 4096
Restart        = always
RestartSec     = 5s
WorkingDirectory=/opt/pocketbase
ExecStart=/opt/pocketbase/pocketbase serve --http=127.0.0.1:8090
UMask=0077

[Install]
WantedBy=multi-user.target
```

I havent change anything this week. 

What could be the issue? How can I analyse out of memory? What memory? there is tooo much enough space. Am I missing anything?

**Comment by ganigeorgiev** [comment]
No one will be able to help you without some steps to reproduce.

In any case, is this a stock PocketBase executable or you have custom build?

If the latter:
- What is this custom code doing?
- Did you build the executabale with `CGO_ENABLED=0` or `CGO_ENABLED=1` (see https://github.com/pocketbase/pocketbase/discussions/4624)?

> data.db and logs.db are on the SSD, storage is on the HDD.

Are you sure that you've mounted the directories correctly (_again, see the above linked issue as it could be also permissions related_)?

>  but now it fails to generate thumbs, the logs show ...

I don't think you are reporting the complete log. Note that you can also find the warning error in the _Admin UI > Logs_ section.
This error usually could be returned if the thumbs generation takes too long to complete (> 60sec; this timeout is not configurable at the moment), e.g. generating thumb of 10MB+ image.

**Reply by ganigeorgiev** [reply]
Also, do you really need `UMask=0077` in your service file?

**Reply by boldeuler** [reply]
Thank you for your reply, it is not reproducable. I have more than 5 million images. 

It is custom build with CGO=0, because I am using OnFileDownloadRequest hook to apply some image manipulations.

I raised the LimitNOFILE    = 4096 to 65536, seems to be working, but I don't understand why.

It took ~2 minutes to restart after raising the LimitNOFILE. Now it does generate thumbs and everything looks right. It has been 20 minutes.

The images are around 6T and yes they are on the HDD.

So,  for the last 20 minutes, it does deliver images, it does generate thumbs, it does apply manipulations. 

I receive 50k - 80k get requests per hour.

**Reply by boldeuler** [reply]
UMask=0077 has its own story. I think it is unnecessary. I will remove it.

**Reply by boldeuler** [reply]
I am still facing some weird issues, for some images it works very well, and some other even the 100x100 thumb runs to 6 minutes timeout, without any logs. After almost 1 hour I started to see few logs.

```
Fallback to original - failed to create thumb 0x2160_ar6_l2964_gog6Y2psek.JPG
```

If I upload an image right now, it works, and it does generate thumb on request.


PS. I have the following thumbs; 

```
0x100, 0x144, 0x240, 0x360, 0x480, 0x720, 0x1080, 0x1440, 0x2160 
```

**Reply by ganigeorgiev** [reply]
You are not reporting the full error message and I don't know how to help you (your refusal to elaborate on the details, how your custom code works and what actually is your application doing doesn't help narrowing the issue).

Again, I suspect that the above thumb error is caused by the thumbs generation timeout. Currently all thumbs are restricted to max 60s and the thumbs workers are limited to `runtime.NumCPU() + 2` (_this is because of https://github.com/pocketbase/pocketbase/pull/3794_). This limit can be eventually made configurable but I need to understand the problem, to somehow replicate it before doing any changes because it could have side-effects and cause issues for other users.

If you want to experiment, clone the repo and change the code in https://github.com/pocketbase/pocketbase/blob/master/apis/file.go.

More than this I don't know how to help further.

_As a side-note, I'll consider with the upcoming v0.23.0 to make the thumbs generation limits at least configurable with ENV variables to make experimentation like this easier._

**Reply by boldeuler** [reply]
I noticed something weird. It is not pocketbase problem, some collections (directories) stopped working. Doing

```
cd home/storage/pocketbase/collections/collectionA/recordA (collection A has over 2m directories (works)
```
and
```
home/storage/pocketbase/collections/collectionB/recordB (collection B has 200k directories (doesnt work)
```

All collections contain only images.


What would make records unreachable? The terminal just hangs.

Trying many operations such as ls -l on the record directory hangs the terminal. 
This is also the reason why pocketbase is not able to open the file (timeout after 6 minutes)

Does it tell you anything?

**Reply by ganigeorgiev** [reply]
No sorry, it could be anything - from hardware/disk error, OS misconfiguration to PocketBase bug.

You are feeding me with pieces of random information that doesn't really make sense because I have literally no context about your project and I cannot do anything else than Google search it for you. I'm not here to provide dedicated support, nor to debug your environment. And even if I wanted to you refuse to provide any details about your application or the actual error that I've previously requested so I really don't know what are the expectations here.  This is not a general dev support forum. If you think that this is a PocketBase issue, elaborate more clearly what triggers it so that I can try to replicate it, otherwise we are both just wasting our time.
